qiskit/utils/lazy_tester.py

```diff
@@ -0,0 +1,276 @@
+
+        @HAS_MATPLOTLIB.require_in_call
+        def my_visualisation():
+            ...
+
+        def my_other_visualisation():
+            # ... some setup ...
+            HAS_MATPLOTLIB.require_now("my_other_visualisation")
+            ...
+
+        def my_third_visualisation():
+            if HAS_MATPLOTLIB:
+                from matplotlib import pyplot
+            else:
+                ...
+
+    In all of these cases, ``matplotlib`` is not imported until the functions are entered.  In the
+    case of the decorator, ``matplotlib`` is tested for import when the function is called for
+    the first time.  In the second and third cases, the loader attempts to import ``matplotlib``
+    when the :meth:`require_now` method is called, or when the Boolean context is evaluated.  For
+    the ``require`` methods, an error is raised if the library is not available.
+
+    This is the base class, which provides the Boolean context checking and error management.  The
+    concrete classes :class:`LazyImportTester` and :class:`LazySubprocessTester` provide convenient
+    entry points for testing that certain symbols are importable from modules, or certain
+    command-line tools are available, respectively.
+    """
+
+    __slots__ = ("_bool", "_callback", "_name", "_install", "_msg")
+
+    def __init__(self, *, name=None, callback=None, install=None, msg=None):
+        """
+        Args:
+            name: the name of this optional dependency.
+            callback: a callback that is called immediately after the availability of the library is
+                tested with the result.  This will only be called once.
+            install: how to install this optional dependency.  Passed to
+                :class:`.MissingOptionalLibraryError` as the ``pip_install`` parameter.
+            msg: an extra message to include in the error raised if this is required.
+        """
+        self._bool = None
+        self._callback = callback
+        self._name = name
+        self._install = install
+        self._msg = msg
+
+    @abc.abstractmethod
+    def _is_available(self) -> bool:
+        """Subclasses of :class:`LazyDependencyManager` should override this method to implement the
+        actual test of availability.  This method should return a Boolean, where ``True`` indicates
+        that the dependency was available.  This method will only ever be called once.
+
+        :meta public:
+        """
+        return False
+
+    def __bool__(self):
+        if self._bool is None:
+            self._bool = self._is_available()
+            if self._callback is not None:
+                self._callback(self._bool)
+        return self._bool
+
+    @typing.overload
+    def require_in_call(self, feature_or_callable: Callable) -> Callable:
+        ...
+
+    @typing.overload
+    def require_in_call(self, feature_or_callable: str) -> Callable[[Callable], Callable]:
+        ...
+
+    def require_in_call(self, feature_or_callable):
```

Comment:
I'm wondering if we should have a class decorator like this too `require_for_instance` that will wrap the class's `__init__` with a require now. Not sure it makes much of a difference, but there is a lot of places where we manually call `require_now()` in the `__init__` but before the super and I'm wondering if having a simpler pattern like
```
@_optionals.HAS_Z3.require_for_instance
class FancyTranspilerPass(TransformationPass):
```

might be less error prone. Although it's been a while since I wrote a class decorator so it might not be trivial.